src/interpreter.cr

```diff
@@ -0,0 +1,95 @@
+
+struct Interpreter
+  @@shell : String | Nil
+
+  protected def self.shell : String
+    {% if flag?(:darwin) %}
+      @@shell ||= begin
+        Process.find_executable(ENV.fetch("RUBYSHELL", "dash")).as(String)
+      rescue TypeCastError
+        Process.find_executable(ENV.fetch("RUBYSHELL", "sh")).as(String)
+      end
+    {% else %}
+      @@shell ||= Process.find_executable(ENV.fetch("RUBYSHELL", "sh")).as(String)
+    {% end %}
+  end
+
+  # Ripped from Crystal::System::FileDescriptor#pipe, except we don't want `reader` to close-on-exec.
+  protected def self.pipe(command : String) : {IO::FileDescriptor, IO::FileDescriptor}
+    pipe_fds = uninitialized StaticArray(LibC::Int, 2)
+    raise IO::Error.from_errno("Could not create pipe") if LibC.pipe(pipe_fds) != 0
+
+    reader = IO::FileDescriptor.new(pipe_fds[0], false)
+    writer = IO::FileDescriptor.new(pipe_fds[1], false)
+    writer.close_on_exec = true
+    writer.sync = true
+    writer << command
+
+    {reader, writer}
+  end
+
+  # Ditto, except we only yield the path.
+  protected def self.pipe(command : String, &)
+    reader, writer = pipe(command)
+
+    begin
+      yield "/dev/fd/#{reader.fd}"
+    ensure
+      writer.flush
+      reader.close
+      writer.close
+    end
+  end
+
+  #######################################################################################
+
+  getter :name, :options, :script
+
+  @name : String = shell
+  @options : Array(String) = [] of String
+  @script : String = ""
+
+  def initialize(content : String)
```

Comment:
Is there a more specific name than `content` for what this argument contains?

Reply:
Yeah, maybe `script` would be a better name for this.

Reply:
Yeah, that's good.